File: app/src/main/java/com/waifusims/wanicchou/widgets/WordListAdapter.java
package com.waifusims.wanicchou.widgets;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.waifusims.wanicchou.R;

import java.util.List;

import data.vocab.WordListEntry;
import data.vocab.models.DictionaryType;

/**
 * RecyclerView adapter for related words
 * Created by Limegrass on 4/4/2018.
 */
public class WordListAdapter<VH extends RecyclerView.ViewHolder>
        extends SelectableAdapter<WordListAdapter.WordViewHolder> {
    private static final String TAG = WordListAdapter.class.getSimpleName();

    private WordViewHolder.ListItemClickListener mListItemClickListener;
    private List<WordListEntry> mWordList;

    public WordListAdapter(List<WordListEntry> words, WordViewHolder.ListItemClickListener listener){
        mWordList = words;
        mListItemClickListener = listener;
    }

    @Override
    public WordViewHolder onCreateViewHolder(ViewGroup parent, int viewType){
        Context context = parent.getContext();
        int wordLayoutId = R.layout.word_list_item;
        LayoutInflater inflater = LayoutInflater.from(context);
        View view = inflater.inflate(wordLayoutId, parent, false);
        return new WordViewHolder(view, mListItemClickListener);
    }


    @Override
    public void onBindViewHolder(WordViewHolder holder, int position) {
//        Log.d(TAG, "#" + position);
        final WordListEntry word = mWordList.get(position);
        holder.showSelectedOverlay(isSelected(position));
        holder.setDictionaryType(mWordList.get(position).getDictionaryType());
        holder.setRelatedWord(mWordList.get(position).getRelatedWord());
    }

    public String getWordAtPosition(int position){
        return mWordList.get(position).getRelatedWord();
    }

    @Override
    public int getItemCount() {
        return mWordList.size();
    }

    public static class WordViewHolder extends RecyclerView.ViewHolder
            implements View.OnLongClickListener, View.OnClickListener {
        private TextView mTVDictionary;
        private TextView mTVWord;
        private View mViewSelectedOverlay;
        private ListItemClickListener mLongClickListener;

        public WordViewHolder(View view, ListItemClickListener listener){
            super(view);
            mTVDictionary = view.findViewById(R.id.tv_word_dic_type);
            mTVWord = view.findViewById(R.id.tv_word_item);
            mViewSelectedOverlay = view.findViewById(R.id.view_selected_overlay);
            mLongClickListener = listener;
            view.setOnLongClickListener(this);
        }

        public void setDictionaryType(DictionaryType dictionaryType){
            mTVDictionary.setText(dictionaryType.toDisplayText());
        }
        public void setRelatedWord(String word){

            mTVWord.setText(word);
        }

        public void showSelectedOverlay(boolean selected){
            mViewSelectedOverlay.setVisibility(selected ? View.VISIBLE : View.INVISIBLE);
        }

        @Override
        public boolean onLongClick(View view) {
            int position = getAdapterPosition();
            Log.d(TAG, "Word index: " + position);
            mLongClickListener.onItemLongClick(position);
            return true;
        }

        @Override
        public void onClick(View view) {
            int position = getAdapterPosition();
            mLongClickListener.onItemClick(position);
        }

        public interface ListItemClickListener {
            void onItemLongClick(int position);
            void onItemClick(int position);
        }
    }
}

File: app/src/main/java/com/waifusims/wanicchou/widgets/WordListAdapter.kt
//package com.waifusims.wanicchou.widgets
//
//import android.support.v7.widget.RecyclerView
//import android.util.Log
//import android.view.LayoutInflater
//import android.view.View
//import android.view.ViewGroup
//import android.widget.TextView
//
//import com.waifusims.wanicchou.R
//
//import data.vocab.shared.WordListEntry
//
///**
// * RecyclerView adapter for related words
// * Created by Limegrass on 4/4/2018.
// */
//class WordListAdapter<VH : RecyclerView.ViewHolder>(private val mWordList: List<WordListEntry>, private val mListItemClickListener: WordViewHolder.ListItemClickListener) : SelectableAdapter<WordListAdapter.WordViewHolder>() {
////
////    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WordViewHolder {
////        val context = parent.context
////        val wordLayoutId = R.layout.word_list_item
////        val inflater = LayoutInflater.from(context)
////        val view = inflater.inflate(wordLayoutId, parent, false)
////        return WordViewHolder(view, mListItemClickListener)
////    }
////
////
////    override fun onBindViewHolder(holder: WordViewHolder, position: Int) {
////        //        Log.d(TAG, "#" + position);
////        val word = mWordList[position]
////        holder.showSelectedOverlay(isSelected(position))
////        holder.setDictionaryType(mWordList[position].dictionaryType)
////        holder.setRelatedWord(mWordList[position].relatedWord)
////    }
////
////    fun getWordAtPosition(position: Int): String? {
////        return mWordList[position].relatedWord
////    }
////
////    override fun getItemCount(): Int {
////        return mWordList.size
////    }
////
////    class WordViewHolder(view: View, private val mLongClickListener: ListItemClickListener) : RecyclerView.ViewHolder(view), View.OnLongClickListener, View.OnClickListener {
////        private val mTVDictionary: TextView
////        private val mTVWord: TextView
////        private val mViewSelectedOverlay: View
////
////        init {
////            mTVDictionary = view.findViewById(R.id.tv_word_dic_type)
////            mTVWord = view.findViewById(R.id.tv_word_item)
////            mViewSelectedOverlay = view.findViewById(R.id.view_selected_overlay)
////            view.setOnLongClickListener(this)
////        }
////
////        fun setDictionaryType(dictionaryType: DictionaryType?) {
////            mTVDictionary.text = dictionaryType!!.toDisplayText()
////        }
////
////        fun setRelatedWord(word: String?) {
////
////            mTVWord.text = word
////        }
////
////        fun showSelectedOverlay(selected: Boolean) {
////            mViewSelectedOverlay.visibility = if (selected) View.VISIBLE else View.INVISIBLE
////        }
////
////        override fun onLongClick(view: View): Boolean {
////            val position = adapterPosition
////            Log.d(TAG, "Word index: $position")
////            mLongClickListener.onItemLongClick(position)
////            return true
////        }
////
////        override fun onClick(view: View) {
////            val position = adapterPosition
////            mLongClickListener.onItemClick(position)
////        }
////
////        interface ListItemClickListener {
////            fun onItemLongClick(position: Int)
////            fun onItemClick(position: Int)
////        }
////    }
////
////    companion object {
////        private val TAG = WordListAdapter<*>::class.java!!.getSimpleName()
////    }
//}
